lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Factory.h"
+#include "cudaq/Optimizer/CodeGen/Passes.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_UNITARYSYNTHESIS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "unitary-synthesis"
+
+using namespace mlir;
+
+namespace cudaq {
+std::vector<std::complex<double>>
+readGlobalConstantArray(cudaq::cc::GlobalOp &global);
+}
+
+namespace {
+
+struct EulerAngles {
+  double alpha;
+  double beta;
+  double gamma;
+};
+
+struct BasisZYZ {
+
+  std::array<std::complex<double>, 4> matrix;
+  EulerAngles angles;
+  /// Updates to the global phase
+  double phase;
+
+  /// This logic is based on https://arxiv.org/pdf/quant-ph/9503016 and its
+  /// corresponding explanation in https://threeplusone.com/pubs/on_gates.pdf,
+  /// Section 4.
+  void decompose() {
+    using namespace std::complex_literals;
+
+    /// Rescale the input unitary matrix, `u`, to be special unitary.
+    /// Extract a phase factor, `phase`, so that `det(inv_phase * unitary) = 1`
+    auto det = (matrix[0] * matrix[3]) - (matrix[1] * matrix[2]);
+    phase = 0.5 * std::atan2(det.imag(), det.real());
+    auto inv_phase = std::exp(-1i * phase);
+
+    std::array<std::complex<double>, 4> special_unitary;
+    std::transform(matrix.begin(), matrix.end(), special_unitary.begin(),
+                   [&](auto element) { return element * inv_phase; });
+
+    auto abs_00 = std::abs(special_unitary[0]);
+    auto abs_01 = std::abs(special_unitary[1]);
+
+    if (abs_00 >= abs_01)
+      angles.beta = 2 * std::acos(abs_00);
+    else
+      angles.beta = 2 * std::asin(abs_01);
+
+    auto half_sum_alpha_gamma =
+        std::atan2(special_unitary[3].imag(), special_unitary[3].real());
+    auto half_diff_alpha_gamma =
+        std::atan2(special_unitary[2].imag(), special_unitary[2].real());
+
+    angles.alpha = half_sum_alpha_gamma + half_diff_alpha_gamma;
+    angles.gamma = half_sum_alpha_gamma - half_diff_alpha_gamma;
+  }
+
+  BasisZYZ(std::vector<std::complex<double>> vec) {
+    std::move(vec.begin(), vec.end(), matrix.begin());
```

Comment:
Should we add a check here to make sure that `vec` is of length 4?

Reply:
Addressed in 3e71dee865f261ba1ba9379e172eb4b68ccab2d7